lib/Dialect/Moore/MooreOps.cpp

```diff
@@ -265,6 +266,61 @@ VariableOp::handlePromotionComplete(const MemorySlot &slot, Value defaultValue,
   return std::nullopt;
 }
 
+SmallVector<DestructurableMemorySlot> VariableOp::getDestructurableSlots() {
+  auto refType = getType();
+  auto destructurable = llvm::dyn_cast<DestructurableTypeInterface>(refType);
+  if (!destructurable)
+    return {};
+
+  auto destructuredType = destructurable.getSubelementIndexMap();
+  if (!destructuredType)
+    return {};
+
+  return {DestructurableMemorySlot{{getResult(), refType}, *destructuredType}};
+}
+
+DenseMap<Attribute, MemorySlot> VariableOp::destructure(
+    const DestructurableMemorySlot &slot,
+    const SmallPtrSetImpl<Attribute> &usedIndices, OpBuilder &builder,
+    SmallVectorImpl<DestructurableAllocationOpInterface> &newAllocators) {
+  builder.setInsertionPointAfter(*this);
+
+  DenseMap<Attribute, MemorySlot> slotMap;
+
+  auto destructurable = llvm::cast<DestructurableTypeInterface>(getType());
+  llvm::ArrayRef<StructLikeMember> members;
+  TypeSwitch<Type>(getType().getNestedType())
+      .Case<StructType, UnpackedStructType>(
+          [&members](auto &type) { members = type.getMembers(); })
+      .Default([this](auto &) {
+        emitOpError("Result type must be StructType or UnpackedStructType");
```

Comment:
Why? Could you not handle any destructurable type?

Reply:
If you cannot handle the other cases, you should either:
- Assert it if the verifier already ensures that the only destructurable types are structs.
- Not emit a destructurable slot in the first place if you cannot actually destructure it.

Reply:
Other cases are basic value which will apply PromotableAllocationOpInterface trait.
Only nested types will apply DestructurableAllocationOpInterface trait.
As mentioned, I can be sure structCreatOp only handle for nested types. But @fabianschuiki suggested not.

Reply:
Okay let me phrase it another way: in what circumstances would this check fail? Would that be considered a problem with the user input or with the compiler logic?

Reply:
You are right, that should not emit an error.